keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+- Fixing _all the problems_ our components have right now, and expanding this beyond what's really necessary. Instead working incrementally, and starting to break out some basic stuff we _know_ every component must handle (e.g. configuration and flag parsing)
+
+### Related proposals / references
+
+- [Kubernetes Component Configuration](https://docs.google.com/document/d/1arP4T9Qkp2SovlJZ_y790sBeiWXDO6SG10pZ_UUU-Lc/edit) by [@mikedanese](https://github.com/mikedanese)
+- [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#) by [@mtaufen](https://github.com/mtaufen)
+- [Moving ComponentConfig API types to staging repos](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md) by [@luxas](https://github.com/luxas) and [@sttts](https://github.com/sttts)
+
+## Proposal
+
+This proposal contains three logical units of work. Each subsection is explained in more detail below.
+
+### Part 1: ComponentConfig
+
+#### Standardized encoding/decoding
+
+- Encoding/decoding helper methods in `k8s.io/component/config/serializer` that would be referenced in every scheme package
+- Warn or (if desired, error) on unknown fields
+  - This has the benefit that it makes it possible for the user to spot e.g. config typos. More high-level, this can be used for e.g. a `--validate-config` flag
+- Support both JSON and YAML
+- Support multiple YAML documents if needed
+
+#### Testing helper methods
+
+- Conversion / roundtrip testing
+- API group testing
+  - External types must have JSON tags
+  - Internal types must not have any JSON tags
+  - ...
+- Defaulting testing
+
+#### Generate OpenAPI specifications
+
+Provide a common way to generate OpenAPI specifications local to the component, so that external consumers can access it, and the component can expose it via e.g. a CLI flag or HTTPS endpoint.
+
+### Part 2: Command building / flag parsing
+
+#### Wrapper around cobra.Command
+
+See the `cmd/kubelet` code for how much extra setup a Kubernetes component needs to do for building commands and flag sets. This code can be refactored into a generic wrapper around _cobra_ for use with Kubernetes.
+
+#### Flag precedence over config file
+
+If the component supports both ComponentConfiguration and flags, flags should override fields set in the ComponentConfiguration. This is not straightforward to implement in code, and only the kubelet does this at the moment. Refactoring this code in a generic helper library in this new repository will make adoption of the feature easy and testable. The details of flag versus ComponentConfig semantics are to be decided later in a different proposal. Meanwhile, this flag precedence feature will be opt-in, so the kubelet and kubeadm can directly adopt this code, until the details have been decided on for all components.
+
+#### Standardized logging
```

Comment:
Does this fit under "command building/flag parsing"?
At a glance, it feels orthogonal to a lot of the other things in this doc (aside from being a thing we can standardize).
Is this just something we can do in parallel with everything else?

Reply:
This is in k8s.io/apiserver/pkg/util/logs today, but obviously does not belong there as it has nothing to do with API servers in particular.

Reply:
Agree that it _strictly speaking_ maybe doesn't really belong, but practically it's very related.
As @sttts said, I'd first like to break the Godep hell of today, put it here, and then think more if it should be its own top-level area or whatever.